ksm.go

```diff
@@ -0,0 +1,458 @@
+		return err
+	}
+
+	k.initialSleepInterval, err = k.sleepInterval.read()
+	if err != nil {
+		_ = k.run.close()
+		_ = k.sleepInterval.close()
+		return err
+	}
+
+	k.initialKSMRun, err = k.run.read()
+	if err != nil {
+		_ = k.run.close()
+		_ = k.sleepInterval.close()
+		return err
+	}
+
+	k.initialized = true
+	k.currentKnob = ksmAggressive
+	k.kickChannel = make(chan bool)
+
+	return nil
+}
+
+func (k *ksm) restore() error {
+	var err error
+
+	k.Lock()
+	defer k.Unlock()
+
+	if !k.initialized {
+		return errors.New("KSM is unavailable")
+	}
+
+	if err = k.pagesToScan.write(k.initialPagesToScan); err != nil {
+		return err
+	}
+
+	if err = k.sleepInterval.write(k.initialSleepInterval); err != nil {
+		return err
+	}
+
+	if err = k.run.write(k.initialKSMRun); err != nil {
+		return err
+	}
+
+	if err := k.run.close(); err != nil {
+		return err
+	}
+
+	if err := k.sleepInterval.close(); err != nil {
+		return err
+	}
+
+	if err := k.pagesToScan.close(); err != nil {
+		return err
+	}
+
+	k.initialized = false
+	return nil
+}
+
+func (k *ksm) tune(s ksmSetting) error {
+	k.Lock()
+	defer k.Unlock()
+
+	if !k.initialized {
+		return errors.New("KSM is unavailable")
+	}
+
+	newPagesToScan, err := s.pagesToScan()
+	if err != nil {
+		return err
+	}
+
+	if err = k.pagesToScan.write(newPagesToScan); err != nil {
+		return err
+	}
+
+	if err = k.sleepInterval.write(fmt.Sprintf("%v", s.scanIntervalMS)); err != nil {
+		return err
+	}
+
+	if err = k.run.write(ksmRun); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type ksmThrottleInterval struct {
+	interval time.Duration
+	nextKnob ksmSettingKnob
+}
+
+var ksmThrottleIntervals = map[ksmSettingKnob]ksmThrottleInterval{
+	ksmAggressive: {
+		interval: 30 * time.Second, // 30s before throttling down from aggressive to standard.
+		nextKnob: ksmStandard,
+	},
+
+	ksmStandard: {
+		interval: 120 * time.Second, // 120s before throttling down from standard to slow.
+		nextKnob: ksmSlow,
+	},
+
+	ksmSlow: {
+		interval: 0, // We stop here unless a new container shows up
+	},
+}
+
+func (k *ksm) throttle() {
+	k.Lock()
+	defer k.Unlock()
+
+	if !k.initialized {
+		proxyLog.Error(errors.New("KSM is unavailable"))
+		return
+	}
+
+	go func() {
+		throttleTimer := time.NewTimer(ksmThrottleIntervals[k.currentKnob].interval)
+
+		for {
+			select {
+			case <-k.kickChannel:
+				// We got kicked, this means a new VM has been created.
+				// We will enter the aggressive setting until we throttle down.
+				_ = throttleTimer.Stop()
+				if err := k.tune(ksmSettings[ksmAggressive]); err != nil {
+					proxyLog.Error(err)
+				} else {
+					k.Lock()
+					k.currentKnob = ksmAggressive
+					k.Unlock()
+				}
+
+				_ = throttleTimer.Reset(ksmThrottleIntervals[k.currentKnob].interval)
+
+			case <-throttleTimer.C:
+				// Our throttling down timer kicked in.
+				// We will move down to the next knob and start the next time,
+				// if necessary.
+				if ksmThrottleIntervals[k.currentKnob].interval == 0 {
```

Comment:
Ideally we should move down once the number of KSMed pages holds steady within a margin of error. At least this is true when the workloads are not active. It will be worth measuring how KSM performs under load and check if this property holds true.